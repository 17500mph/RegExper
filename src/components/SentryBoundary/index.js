import React from 'react';
import PropTypes from 'prop-types';
import * as Sentry from '@sentry/browser';

import SentryError from 'components/SentryError';

class SentryBoundary extends React.Component {
  state = {
    hasError: false
  }

  propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ]).isRequired
  }

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  componentDidCatch(error, errorInfo) {
    Sentry.withScope(scope => {
      Object.keys(errorInfo).forEach(key =>
        scope.setExtra(key, errorInfo[key]));
      Sentry.captureException(error);
    });
  }

  render() {
    const { hasError } = this.state;
    const { children } = this.props;

    if (hasError) {
      return <SentryError />;
    } else {
      return children;
    }
  }
}

export default SentryBoundary;
