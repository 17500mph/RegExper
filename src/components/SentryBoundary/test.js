jest.mock('@sentry/browser');

import React from 'react';
import { shallow } from 'enzyme';
import * as Sentry from '@sentry/browser';

import SentryBoundary from 'components/SentryBoundary';

describe('SentryBoundary', () => {
  test('rendering', () => {
    const component = shallow(
      <SentryBoundary>
        Example content
      </SentryBoundary>
    );
    expect(component).toMatchSnapshot();
  });

  test('error handling', () => {
    const Child = () => 'Example content';
    const component = shallow(
      <SentryBoundary>
        <Child />
      </SentryBoundary>
    );
    expect(component).toMatchSnapshot();

    const error = new Error('Example error');
    component.find('Child').simulateError(error);
    // NOTE: Enzyme doesn't call getDerivedStateFromError yet, so we have to
    // set the state manually
    component.setState(SentryBoundary.getDerivedStateFromError(error));

    const scope = { setExtra: jest.fn() };
    expect(Sentry.withScope).toHaveBeenCalled();

    Sentry.withScope.mock.calls[0][0](scope);
    expect(Sentry.captureException).toHaveBeenCalledWith(error);
    expect(component).toMatchSnapshot();
  });
});
