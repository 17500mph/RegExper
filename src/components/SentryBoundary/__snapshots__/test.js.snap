// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SentryBoundary error handling 1`] = `<Child />`;

exports[`SentryBoundary error handling 2`] = `<withI18nextTranslation(SentryError) />`;

exports[`SentryBoundary rendering 1`] = `"Example content"`;
